scopehal/MockOscilloscope.cpp

```diff
@@ -525,3 +528,160 @@ bool MockOscilloscope::LoadCSV(const string& path)
+	f.copy((char*)&fh, sizeof(FileHeader), fpos);
+	fpos += sizeof(FileHeader);
+
+	//Get vendor from file signature
+	switch(fh.magic[0])
+	{
+		case 'A':
+			m_vendor = "Agilent/Keysight";
+			break;
+
+		case 'R':
+			m_vendor = "Rigol";
+			break;
+
+		default:
+			LogError("Unknown file format");
+			return false;
+	}
+	
+	LogDebug("Vendor:    %s\n", m_vendor.c_str());
+	//LogDebug("File size: %i bytes\n", fh.length);
+	LogDebug("Waveforms: %i\n\n", fh.count);
+
+	//Load waveforms
+	for(size_t i=0; i<fh.count; i++)
+	{
+		LogDebug("Waveform %i:\n", (int)i+1);
+		LogIndenter li_w;
+
+		//Parse waveform header
+		WaveHeader wh;
+		f.copy((char*)&wh, sizeof(WaveHeader), fpos);
+		fpos += sizeof(WaveHeader);
+
+		// Only set name/serial on first waveform
+		if (i == 0)
+		{
+			// Split hardware string
+			int idx = 0;
+			for(int c=0; c<24; c++)
+			{
+				if(wh.hardware[c] == ':')
+				{
+					idx = c;
+					break;
+				}
+			}
+
+			//Set oscilloscope metadata
+			m_name.assign(wh.hardware, idx);
+			m_serial.assign(wh.hardware + idx + 1, 24 - idx);
+		}
+
+		LogDebug("Samples:      %i\n", wh.samples);
+		LogDebug("Buffers:      %i\n", wh.buffers);
+		LogDebug("Type:         %i\n", wh.type);
+		LogDebug("Duration:     %.*f us\n", 2, wh.duration * 1e6);
+		LogDebug("Start:        %.*f us\n", 2, wh.start * 1e6);
+		LogDebug("Interval:     %.*f ns\n", 2, wh.interval * 1e9);
+		LogDebug("Origin:       %.*f us\n", 2, wh.origin * 1e6);
+		LogDebug("Holdoff:      %.*f ms\n", 2, wh.holdoff * 1e3);
+		LogDebug("Sample Rate:  %.*f Msps\n", 2, (1 / wh.interval) / 1e6);
+		LogDebug("Frame:        %s\n", m_name.c_str());
+		LogDebug("Serial:       %s\n\n", m_serial.c_str());
+
+		// Create new channel
+		auto chan = new OscilloscopeChannel(
+			this,			//Parent scope
+			wh.label,		//Channel name
+			OscilloscopeChannel::CHANNEL_TYPE_ANALOG,
+			GetDefaultChannelColor(i),
+			units[wh.x],
+			units[wh.y],
+			1,				//Bus width
+			i,				//Channel index
+			true			//Is physical channel
+		);
+		AddChannel(chan);
+		chan->SetDefaultDisplayName();
+
+		//Create new waveform for channel
+		auto wfm = new AnalogWaveform;
+		wfm->m_timescale = wh.interval * 1e15;
+		wfm->m_startTimestamp = 0;
+		wfm->m_startFemtoseconds = 0;
+		wfm->m_triggerPhase = 0;
+		chan->SetData(wfm, 0);
+
+		//Loop through waveform buffers
+		float vmin = FLT_MAX;
+		float vmax = -FLT_MAX;
+		for(size_t j=0; j<wh.buffers; j++)
+		{
+			LogDebug("Buffer %i:\n", (int)j+1);
+			LogIndenter li_b;
+
+			//Parse waveform data header
+			DataHeader dh;
+			f.copy((char*)&dh, sizeof(DataHeader), fpos);
+			fpos += sizeof(DataHeader);
+
+    		LogDebug("Data Type:      %i\n", dh.type);
+    		LogDebug("Sample depth:   %i bits\n", dh.depth*8);
+    		LogDebug("Buffer length:  %i KB\n\n\n", dh.length/1024);
+
+			//Loop through waveform samples
+			float* sample_f = nullptr;
+			uint8_t* sample_i = nullptr;
+			float sample = 0;
+			for(size_t k=0; k<wh.samples; k++)
+			{
+				if (dh.type == 6)
```

Comment:
This code can be optimized to avoid doing comparison in a big loop (which slow down process for nothing)
You can do the comparison before the for() and have 2 branch

Reply:
Good idea. Data type check is now outside the loop.